// lehre/dist/src/parsers/babel/functions.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var functions_1 = require("./functions");
var helper_1 = require("./helper");
describe('functions', function () {
    it('getFunctionDoc', function () {
        var code = "\n    function foo() {\n    }\n    ";
        var ast = helper_1.getAst(code);
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        var actual = functions_1.getFunctionDoc(ast.program.body[0], code.split('\n'));
        expect(actual).toEqual({
            name: 'foo',
            type: 'function',
            start: { line: 1, column: 4 },
            end: { line: 2, column: 5 },
            params: [],
            returnType: '',
        });
    });
    it('getFunctionDoc with args', function () {
        var code = "\n    function foo(arg1: string, arg2: number) {\n    }\n    ";
        var ast = helper_1.getAst(code);
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        var actual = functions_1.getFunctionDoc(ast.program.body[0], code.split('\n'));
        expect(actual).toEqual({
            name: 'foo',
            type: 'function',
            start: { line: 1, column: 4 },
            end: { line: 2, column: 5 },
            params: [
                {
                    name: 'arg1',
                    type: 'string',
                    default: '',
                    alias: 'string',
                },
                {
                    name: 'arg2',
                    type: 'number',
                    default: '',
                    alias: 'number',
                },
            ],
            returnType: '',
        });
    });
    it('getFunctionDoc with return type', function () {
        var code = "\n    function foo(arg1: string, arg2: number): number {\n    }\n    ";
        var ast = helper_1.getAst(code);
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        var actual = functions_1.getFunctionDoc(ast.program.body[0], code.split('\n'));
        expect(actual).toEqual({
            name: 'foo',
            type: 'function',
            start: { line: 1, column: 4 },
            end: { line: 2, column: 5 },
            params: [
                {
                    name: 'arg1',
                    type: 'string',
                    default: '',
                    alias: 'string',
                },
                {
                    name: 'arg2',
                    type: 'number',
                    default: '',
                    alias: 'number',
                },
            ],
            returnType: 'number',
        });
    });
    it('getFunctionDoc with default parameter', function () {
        var code = "\n    function foo(arg1: string = 'foo', arg2: number = 100): number {\n    }\n    ";
        var ast = helper_1.getAst(code);
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        var actual = functions_1.getFunctionDoc(ast.program.body[0], code.split('\n'));
        expect(actual).toEqual({
            name: 'foo',
            type: 'function',
            start: { line: 1, column: 4 },
            end: { line: 2, column: 5 },
            params: [
                {
                    name: 'arg1',
                    type: 'string',
                    default: "'foo'",
                    alias: 'string',
                },
                {
                    name: 'arg2',
                    type: 'number',
                    default: '100',
                    alias: 'number',
                },
            ],
            returnType: 'number',
        });
    });
    it('getFunctionDoc with alias', function () {
        var code = "\n    function foo(arg1: string, arg2: { foo: number }) {\n    }\n    ";
        var ast = helper_1.getAst(code);
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        var actual = functions_1.getFunctionDoc(ast.program.body[0], code.split('\n'));
        expect(actual).toEqual({
            name: 'foo',
            type: 'function',
            start: { line: 1, column: 4 },
            end: { line: 2, column: 5 },
            params: [
                {
                    name: 'arg1',
                    type: 'string',
                    default: '',
                    alias: 'string',
                },
                {
                    name: 'arg2',
                    type: '{ foo: number }',
                    default: '',
                    alias: 'Object',
                },
            ],
            returnType: '',
        });
    });
    it('getFunctionDoc with export', function () {
        var code = "\n    export function foo(arg1: string = 'foo', arg2: number = 100): number {\n    }\n    ";
        var ast = helper_1.getAst(code);
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        var declaration = ast.program.body[0].declaration;
        var actual = functions_1.getFunctionDoc(declaration, code.split('\n'));
        expect(actual).toEqual({
            name: 'foo',
            type: 'function',
            start: { line: 1, column: 11 },
            end: { line: 2, column: 5 },
            params: [
                {
                    name: 'arg1',
                    type: 'string',
                    default: "'foo'",
                    alias: 'string',
                },
                {
                    name: 'arg2',
                    type: 'number',
                    default: '100',
                    alias: 'number',
                },
            ],
            returnType: 'number',
        });
    });
    it('getFunctionDoc with export default', function () {
        var code = "\n    export default function foo(arg1: string = 'foo', arg2: number = 100): number {\n    }\n    ";
        var ast = helper_1.getAst(code);
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        var declaration = ast.program.body[0].declaration;
        var actual = functions_1.getFunctionDoc(declaration, code.split('\n'));
        expect(actual).toEqual({
            name: 'foo',
            type: 'function',
            start: { line: 1, column: 19 },
            end: { line: 2, column: 5 },
            params: [
                {
                    name: 'arg1',
                    type: 'string',
                    default: "'foo'",
                    alias: 'string',
                },
                {
                    name: 'arg2',
                    type: 'number',
                    default: '100',
                    alias: 'number',
                },
            ],
            returnType: 'number',
        });
    });
});
