// lehre/dist/src/parsers/ts/expressions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getVariableDocFromExpression = void 0;
var parameters_1 = require("./parameters");
var helpers_1 = require("../helpers");
var getVariableDocFromExpression = function (node, source) {
    var params = [];
    var start = source.getLineAndCharacterOfPosition(node.getStart(source));
    var end = source.getLineAndCharacterOfPosition(node.getEnd());
    var doc = {
        name: '',
        type: 'function',
        start: { line: start.line, column: start.character },
        end: { line: end.line, column: end.character },
        returnType: '',
        params: params,
    };
    var _a = node, left = _a.left, right = _a.right;
    doc.name = left.getText(source);
    var func = right;
    doc.params = func.parameters.map(function (p) { return parameters_1.getParameter(p, source); });
    if (helpers_1.has(right, 'type')) {
        doc.returnType = func.type ? func.type.getText(source) : '';
    }
    return doc;
};
exports.getVariableDocFromExpression = getVariableDocFromExpression;
