// lehre/dist/src/parsers/ts/functions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFunctionDoc = void 0;
var parameters_1 = require("./parameters");
var helpers_1 = require("../helpers");
var getFunctionDoc = function (node, source) {
    var params = [];
    var start = source.getLineAndCharacterOfPosition(node.getStart(source));
    var end = source.getLineAndCharacterOfPosition(node.getEnd());
    var doc = {
        name: node.name ? node.name.text : '',
        type: 'function',
        start: { line: start.line, column: start.character },
        end: { line: end.line, column: end.character },
        returnType: '',
        params: params,
    };
    if (helpers_1.has(node, 'parameters')) {
        doc.params = node.parameters.map(function (n) {
            return parameters_1.getParameter(n, source);
        });
    }
    if (helpers_1.has(node, 'type') && node.type) {
        doc.returnType = node.type.getText(source);
    }
    return doc;
};
exports.getFunctionDoc = getFunctionDoc;
