// lehre/dist/src/parsers/ts/types.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var typescript_1 = require("typescript");
var types_1 = require("./types");
var getAcual = function (code) {
    var source = typescript_1.createSourceFile('lehre.ts', code, typescript_1.ScriptTarget.ESNext, false, typescript_1.ScriptKind.TS);
    var actual = types_1.getTypeDoc(source.statements[0], source);
    return actual;
};
describe('types', function () {
    it('getTypeDoc', function () {
        var code = "\n    type Foo = {\n    }";
        var actual = getAcual(code);
        expect(actual).toEqual({
            name: 'Foo',
            type: 'function',
            start: { line: 1, column: 4 },
            end: { line: 2, column: 5 },
            methods: [],
            heritageClauses: [],
        });
    });
    it('getTypeDoc - with props', function () {
        var code = "\n    type Foo = {\n      foo: string\n      bar: () => void\n    }";
        var actual = getAcual(code);
        expect(actual).toEqual({
            name: 'Foo',
            type: 'function',
            start: { line: 1, column: 4 },
            end: { line: 4, column: 5 },
            methods: [],
            heritageClauses: [],
        });
    });
});
