// lehre/dist/src/parsers/ts/variables.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var typescript_1 = require("typescript");
var variables_1 = require("./variables");
describe('variables', function () {
    it('getVariableDoc', function () {
        var code = "const foo = () => {}";
        var source = typescript_1.createSourceFile('lehre.ts', code, typescript_1.ScriptTarget.ESNext, false, typescript_1.ScriptKind.TS);
        var start = { line: 0, column: 0 };
        var end = { line: 0, column: 20 };
        var statement = source.statements[0];
        var actual = variables_1.getVariableDoc(statement.declarationList.declarations[0], source, start, end);
        expect(actual).toEqual({
            name: 'foo',
            type: 'function',
            start: start,
            end: end,
            params: [],
            returnType: '',
        });
    });
    it('getVariableDoc with args', function () {
        var code = "const foo = (arg1: string, arg2: number) => {}\n    ";
        var source = typescript_1.createSourceFile('lehre.ts', code, typescript_1.ScriptTarget.ESNext, false, typescript_1.ScriptKind.TS);
        var start = { line: 0, column: 0 };
        var end = { line: 0, column: 43 };
        var statement = source.statements[0];
        var actual = variables_1.getVariableDoc(statement.declarationList.declarations[0], source, start, end);
        expect(actual).toEqual({
            name: 'foo',
            type: 'function',
            start: start,
            end: end,
            params: [
                {
                    name: 'arg1',
                    type: 'string',
                    default: '',
                    alias: 'string',
                },
                {
                    name: 'arg2',
                    type: 'number',
                    default: '',
                    alias: 'number',
                },
            ],
            returnType: '',
        });
    });
    it('getFunctionDoc with return type', function () {
        var code = "const foo = (arg1: string, arg2: number): number => {}";
        var source = typescript_1.createSourceFile('lehre.ts', code, typescript_1.ScriptTarget.ESNext, false, typescript_1.ScriptKind.TS);
        var start = { line: 0, column: 0 };
        var end = { line: 0, column: 54 };
        var statement = source.statements[0];
        var actual = variables_1.getVariableDoc(statement.declarationList.declarations[0], source, start, end);
        expect(actual).toEqual({
            name: 'foo',
            type: 'function',
            start: start,
            end: end,
            params: [
                {
                    name: 'arg1',
                    type: 'string',
                    default: '',
                    alias: 'string',
                },
                {
                    name: 'arg2',
                    type: 'number',
                    default: '',
                    alias: 'number',
                },
            ],
            returnType: 'number',
        });
    });
    it('getVariableDoc with default parameter', function () {
        var code = "const foo = (arg1: string = 'foo', arg2: number = 100): number => {}";
        var source = typescript_1.createSourceFile('lehre.ts', code, typescript_1.ScriptTarget.ESNext, false, typescript_1.ScriptKind.TS);
        var start = { line: 0, column: 0 };
        var end = { line: 0, column: 65 };
        var statement = source.statements[0];
        var actual = variables_1.getVariableDoc(statement.declarationList.declarations[0], source, start, end);
        expect(actual).toEqual({
            name: 'foo',
            type: 'function',
            start: start,
            end: end,
            params: [
                {
                    name: 'arg1',
                    type: 'string',
                    default: "'foo'",
                    alias: 'string',
                },
                {
                    name: 'arg2',
                    type: 'number',
                    default: '100',
                    alias: 'number',
                },
            ],
            returnType: 'number',
        });
    });
    it('getVariableDoc with alias', function () {
        var code = "const foo = (arg1: string, arg2: { foo: number }) => {}";
        var source = typescript_1.createSourceFile('lehre.ts', code, typescript_1.ScriptTarget.ESNext, false, typescript_1.ScriptKind.TS);
        var start = { line: 0, column: 0 };
        var end = { line: 0, column: 55 };
        var statement = source.statements[0];
        var actual = variables_1.getVariableDoc(statement.declarationList.declarations[0], source, start, end);
        expect(actual).toEqual({
            name: 'foo',
            type: 'function',
            start: start,
            end: end,
            params: [
                {
                    name: 'arg1',
                    type: 'string',
                    default: '',
                    alias: 'string',
                },
                {
                    name: 'arg2',
                    type: '{ foo: number }',
                    default: '',
                    alias: 'Object',
                },
            ],
            returnType: '',
        });
    });
});
