// lehre/dist/src/parsers/ts/variables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getVariableDoc = void 0;
var parameters_1 = require("./parameters");
var helpers_1 = require("../helpers");
var getVariableDoc = function (node, source, start, end) {
    var params = [];
    var doc = {
        name: node.name.text,
        type: '',
        start: start,
        end: end,
        returnType: '',
        params: params,
    };
    if (helpers_1.has(node, 'initializer') &&
        node.initializer &&
        helpers_1.has(node.initializer, 'parameters')) {
        doc.type = 'function';
        var parameters = node.initializer.parameters;
        doc.params = parameters.map(function (param) {
            return parameters_1.getParameter(param, source);
        });
        if (helpers_1.has(node.initializer, 'type')) {
            var type = node.initializer.type;
            if (type) {
                doc.returnType = type.getText(source);
            }
        }
    }
    else {
        if (node.initializer && helpers_1.has(node, 'type') && node.type) {
            doc.type = 'property';
            var type = node.initializer.type;
            if (type) {
                doc.returnType = type.getText(source);
            }
        }
    }
    return doc;
};
exports.getVariableDoc = getVariableDoc;
